### ISSUE TITLE
Notification component

### ISSUE THREAD

**Issue by artalar** [ariakit/ariakit#464]
### Motivation

Notification - is a standard pattern to tell a user about some unimportant event, without blocking the interface.

### Example

https://ant.design/components/message/

### Possible implementations

https://inclusive-components.design/notifications/

**Comment by artalar** [comment]
More info: https://github.com/jackbsteinberg/std-toast/

**Comment by diegohaz** [comment]
My idea for this component is a generic `Alert` component following the respective [WAI-ARIA pattern](https://www.w3.org/TR/wai-aria-practices/#alert).

**Comment by artalar** [comment]
Note: here https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/alertdialog.html#ex_label if you change the text and click _save_ you can see _notification component_  🤔 
![example](https://user-images.githubusercontent.com/27290320/70968164-8455d180-20a9-11ea-9fab-86a14b19d451.png)

**Comment by diegohaz** [comment]
Cool! That's a good example! That element is using `role="alert"` too:
![image](https://user-images.githubusercontent.com/3068563/70986998-d0027e00-209d-11ea-8e86-483bb88853aa.png)

**Comment by artalar** [comment]
@diegohaz have you seen https://inclusive-components.design/notifications/ and https://github.com/jackbsteinberg/std-toast/?

**Comment by folz** [comment]
As I'm working on something related to this - @diegohaz do you have thoughts on how opinionated reakit wants to be? That is, do you want to provide a Notification/Toast/Message-style component like the OP is asking about, or do you want people to build their own Notification component with tools that reakit provides?

Some thoughts:

At one end, you could have something like

```tsx
useAlert = createHook({
  name: 'Alert',
  compose: useBox,
  useProps() { return {role: 'alert', ...htmlProps} }
};
Alert = createComponent({as: 'div', useHook: useAlert});
```
With usage like:
```tsx
<Alert>This is an alert.</Alert>
```

But this has a couple issues. When server rendering, the alert won't be read out because it's part of the page originally. It also feels like an unnecessary layer over adding `role: 'alert'` to each alert-able component.

We could also make Alert compose Hidden:

```tsx
useAlert = createHook({
  name: 'Alert',
  compose: useHidden,
  useState: useAlertState, // calls useHiddenState
  useProps() { return {role: 'alert', ...htmlProps} }
};
Alert = createComponent({as: 'div', useHook: useAlert});
```
With usage like:
```tsx
const alert = useAlertState();
React.useEffect(() => { alert.show(); });
return <Alert {...alert}>This is an alert.</Alert>
```

Or maybe even build in the `alert.show()` effect to the component so it will always cause itself to be read out? That behavior feels like it's more opinionated than what you want reakit to provide, though. Like, will all Alerts really need to show themselves one render pass later? Would providing a boolean option for that behavior introduce unnecessary complexity? (Probably.)

Assuming a `role="alert"` is also an issue because it causes an `assertive` readout. I feel like most Notifications are informational and should carry `role: 'status'` instead so the readout is `polite`. Maybe the component should be an abstract LiveRegion instead? Or should reakit provide multiple abstract live region components? I don't think so, because live regions are fairly straightforward to use, the real challenge is in using them correctly and that can vary based on what the end product goal is.

On the other hand, if reakit wants to provide a Notification component, would it also handle the positioning and ordering of Notifications similar to other Toast solutions? At the very least it should handle persisting Notifications even if the view which created the notification is unmounted. I think this needs a ReakitProvider-level context to manage, which I don't think any other reakit component requires.

Would a Notification component include intent like success/info/error? There's some precedent for this (with FormMessage). We could then do something like setting role: status on `success`/`info`, but role: alert on `error`.

Anyway, these are just some ideas. Would appreciate any thoughts you all have on the matter.

**Comment by diegohaz** [comment]
@folz Thanks for the thoughtful comment! :)

Off the top of my head, I remember seeing a notification pattern where you would have a single live region element on the page (or one for each kind of `aria-live` region - `polite`, `assertive` etc.) and then add/remove messages there as they're dispatched:
```html
<body>
  <div id="root"></div>
  <div aria-live="assertive" aria-atomic="true">
    <!-- New messages are added here -->
    <p>Toast 1</p>
    <!-- What if they have actions? -->
    <p>Toast 2 <button>x</button></p>
    ...
  </div>
</body>
```

If I recall correctly, the main reason for doing that was so subsequent messages would not interrupt the one that is being announced.

That's just an idea. I don't know if that's the best approach, and I'm not sure how that could be done in the "Reakit way".

Regarding `FormMessage`, I'm still not sure if that should be `role="alert"` or `aria-live="polite"`. That's one of the reasons why the Form module is still unstable. That needs more research since there's no recommended pattern on forms on the WAI-ARIA spec.

**Comment by folz** [comment]
Happy new year 🎉 After some exploratory work and talking with my team, we think it makes the most sense for our use case to implement the [Snackbar](https://material.io/components/snackbars/#) pattern.

My read is that this PR is more interested in Growl-style notifications but there are a fair number of accessibility concerns around Snackbars, so it might still be worthwhile to have. I can open a PR once the implementation is ready for review if you like.

**Comment by gziolo** [comment]
There is `Snackbar` component in WordPress components and it’s implemented as a div with the role button:

https://github.com/WordPress/gutenberg/blob/34253056f822cf554437d5ad449b033d811e199f/packages/components/src/snackbar/index.js#L78

As Diego mentioned, the content is announced from the single live region using `speak` function from `@wordpress/a11y` package.

**Comment by beaussan** [comment]
Any update on this ?

**Comment by R-Oscar** [comment]
Hey there!

I would agree with @folz that `role: status` is more suitable for Notification component than `role: alert`. Although there is a suggestion in an [research](https://github.com/jackbsteinberg/std-toast) @artalar provided:

> By default, "error" toasts will be treated as having the ARIA role semantics of alert, while the other toasts will be treated as having the ARIA role semantics of a status.

Maybe we should go after this? We could include two live regions, one for error notifications, and `polite` for others. Also maybe user would like to have the ability to customize `role` by himself?

Anyway, if noone objects, I would dive into it, or provide assistance in @folz PR if he’s willing to do it :)

**Comment by artalar** [comment]
@diegohaz do you follow the thread?  👀

**Comment by diegohaz** [comment]
@artalar Yeah! I don't have time to dig into this myself right now. But, please, keep going.

**Comment by strass** [comment]
I have this PR but I made it to try to figure out Reakit and I'm not sure it's actually worth adding. No response from Haz though... https://github.com/reakit/reakit/pull/787

**Comment by artalar** [comment]
Check this out! https://web.dev/building-a-toast-component/

**Comment by diegohaz** [comment]
Just leaving this exploration from my notes:

---

# Simple notifications

```tsx
import * as Ariakit from "@ariakit/react";

function Example() {
  const notification = Ariakit.useNotificationContext();
  return (
    <>
      <button
        onClick={() => {
          notification?.push("Hello");
          notification?.push("Hello", { role: "status" });
          notification?.push("Hello", { timeout: 10000 });
          notification?.push("Hello", { visuallyHidden: true });
        }}
      >
        Push notification
      </button>
    </>
  );
}

export default function App() {
  return (
    <Ariakit.NotificationProvider>
      <Example />
      <div>
        <Ariakit.NotificationItems>
          {(items) => (
            items.map((props) => (
              <Ariakit.Notification key={props.id} {...props}>
                {props.children}
                <Ariakit.NotificationDismiss />
              </Ariakit.Notification>
            ))
          )}
        </Ariakit.NotificationItems>
      </div>
    </Ariakit.NotificationProvider>
  );
}
```

---

# Custom notifications

```tsx
// notifications.tsx
import * as Ariakit from "@ariakit/react";

interface Item extends Ariakit.NotificationStoreItem {
  onUndo?: () => void;
}

export function useNotification() {
  const notification = Ariakit.useNotificationContext();
  if (!notification) {
    throw new Error("Missing NotificationProvider");
  }
  return notification as Ariakit.NotificationStore<Item>;
}

export function Notifications({ children }) {
  return (
    <Ariakit.NotificationProvider>
      {children}
      <div>
        <Ariakit.NotificationItems<Item>>
          {(items) => (
            items.map(({ onUndo, ...props }) => (
              <Ariakit.Notification key={props.id} {...props}>
                {props.children}
                <Ariakit.NotificationDismiss />
                {onUndo && <button onClick={onUndo}>Undo</button>}
              </Ariakit.Notification>
            ))
          )}
        </Ariakit.NotificationItems>
      </div>
    </Ariakit.NotificationProvider>
  );
}
```

```tsx
// index.tsx
import { Notifications, useNotification } from "./notifications.tsx";

function Example() {
  const { push } = useNotification();
  return (
    <>
      <button
        onClick={() => {
          push("Hello");
          push("Hello", { role: "status" });
          push("Hello", { onUndo: () => {}, timeout: 10000 });
          push("Hello", { visuallyHidden: true });
        }}
      >
        Push notification
      </button>
    </>
  );
}

export default function App() {
  return (
    <Notifications>
      <Example />
    </Notifications>
  );
}
```

---

Comment by Alvar Lagerlöf for future reference:

> How does your notification store handle multiple SSR renders on a single server? Does it leak state considering it's module scoped?
>
> — https://x.com/alvarlagerlof/status/1848740406365753802

This could become an issue if someone needs to render an initial notification on the server. They might have to instantiate the store within the component, which would prevent them from exporting `push`, or they might need to mutate the store on the server, potentially affecting other clients (this needs testing).

Edit: Updated the code.
